### PR TITLE
Rework Plugin System

### PR DESCRIPTION
**Pull request by Jotschi**

### CODE REVIEW

File: api/src/main/java/com/gentics/mesh/Mesh.java [hunk 1]
```diff
@@ -156,4 +159,21 @@ static boolean isVertxReady() {
...
+	 * @param id
+	 *            The id of the plugin (e.g. hello-world)
+	 * @return
+	 */
+	Completable deployPlugin(Class<?> clazz, String id);
+
+	/**
+	 * Return a map of deployed plugins with name and id.
+	 * 
+	 * @return
```
Comment (Jotschi) [inline api/src/main/java/com/gentics/mesh/Mesh.java]: Renamed this

File: api/src/main/java/com/gentics/mesh/plugin/PluginManifest.java [hunk 2]
```diff
@@ -13,6 +8,10 @@
  */
 public class PluginManifest {
 
+	@JsonProperty(required = true)
+	@JsonPropertyDescription("Unique id of the plugin")
```
Comment (Jotschi) [inline api/src/main/java/com/gentics/mesh/plugin/PluginManifest.java]: Origin of id

**Comment by philippguertler** [comment]
Rewrite free plugins section

**Comment by philippguertler** [comment]
Write docs for
* Plugin configuration
* Plugin storage
* Admin client usage
* User client usage
* Starting Mesh from Java main method or test with plugin

**Comment by philippguertler** [comment]
Got this error when starting the plugin the second time:
```
15:59:36.520 [null] ERROR [main] [AbstractPluginManager.java:230] - Cannot find the manifest path
org.pf4j.PluginRuntimeException: Cannot find the manifest path
	at com.gentics.mesh.plugin.impl.MeshPluginDescriptorFinderImpl.readManifest(MeshPluginDescriptorFinderImpl.java:61)
	at com.gentics.mesh.plugin.impl.MeshPluginDescriptorFinderImpl.find(MeshPluginDescriptorFinderImpl.java:43)
	at com.gentics.mesh.plugin.impl.MeshPluginDescriptorFinderImpl.find(MeshPluginDescriptorFinderImpl.java:21)
	at org.pf4j.AbstractPluginManager.loadPluginFromPath(AbstractPluginManager.java:799)
	at com.gentics.mesh.plugin.manager.MeshPluginManagerImpl.loadPluginFromPath(MeshPluginManagerImpl.java:470)
	at org.pf4j.AbstractPluginManager.loadPlugins(AbstractPluginManager.java:228)
	at com.gentics.mesh.plugin.manager.MeshPluginManagerImpl.lambda$deployExistingPluginFiles$0(MeshPluginManagerImpl.java:118)
	at io.reactivex.internal.operators.completable.CompletableDefer.subscribeActual(CompletableDefer.java:36)
	at io.reactivex.Completable.subscribe(Completable.java:1635)
	at io.reactivex.Completable.subscribe(Completable.java:1701)
	at com.gentics.mesh.cli.BootstrapInitializerImpl.init(BootstrapInitializerImpl.java:310)
	at com.gentics.mesh.cli.MeshImpl.run(MeshImpl.java:129)
	at PluginTestRunner.run(PluginTestRunner.java:18)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMainV2.main(AppMainV2.java:131)
```

File: plugin-api/src/main/java/com/gentics/mesh/plugin/GraphQLPlugin.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+	GraphQLObjectType createRootType();
+
+	/**
+	 * Prefix the type of the plugin with the plugin id to avoid conflicts with Gentics Mesh or other plugin types.
+	 * 
+	 * @param type
+	 * @return Prefixed type
+	 */
+	default String prefixType(String type) {
+		// TODO sanitize id
```
Comment (Jotschi) [inline plugin-api/src/main/java/com/gentics/mesh/plugin/GraphQLPlugin.java]: Sanitize Id?

**Comment by philippguertler** [comment]
Add troubleshoot section to the plugin for common problems:

This happens when the constructor is missing
```
16:03:28.243 [null] ERROR [main] [MeshPluginFactory.java:63] - com.gentics.mesh.plugin.PgPlugin.<init>(org.pf4j.PluginWrapper, com.gentics.mesh.plugin.env.PluginEnvironment)
java.lang.NoSuchMethodException: com.gentics.mesh.plugin.PgPlugin.<init>(org.pf4j.PluginWrapper, com.gentics.mesh.plugin.env.PluginEnvironment)
	at java.lang.Class.getConstructor0(Class.java:3082)
	at java.lang.Class.getConstructor(Class.java:1825)
	at com.gentics.mesh.plugin.pf4j.MeshPluginFactory.create(MeshPluginFactory.java:60)
	at org.pf4j.PluginWrapper.getPlugin(PluginWrapper.java:81)
	at com.gentics.mesh.plugin.manager.MeshPluginManagerImpl.lambda$deploy$4(MeshPluginManagerImpl.java:377)
	at io.reactivex.internal.operators.completable.CompletableDefer.subscribeActual(CompletableDefer.java:36)
	at io.reactivex.Completable.subscribe(Completable.java:1635)
	at io.reactivex.Completable.subscribe(Completable.java:1701)
	at PluginTestRunner.run(PluginTestRunner.java:23)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMainV2.main(AppMainV2.java:131)
```

File: plugin-api/src/main/java/com/gentics/mesh/plugin/util/PluginUtils.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+			throw error(BAD_REQUEST, "admin_plugin_error_validation_failed_field_missing", "version");
+		}
+	}
+
+	public static void validate(RestPlugin plugin) {
+		String apiName = plugin.apiName();
+		if (StringUtils.isEmpty(apiName)) {
+			throw error(BAD_REQUEST, "admin_plugin_error_validation_failed_field_missing", "apiName");
+		}
+		if (apiName.contains(" ") || apiName.contains("/")) {
```
Comment (Jotschi) [inline plugin-api/src/main/java/com/gentics/mesh/plugin/util/PluginUtils.java]: Check apiName via regular expression

File: plugin-api/src/main/java/com/gentics/mesh/plugin/util/PluginUtils.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+			throw error(BAD_REQUEST, "admin_plugin_error_validation_failed_apiname_invalid", plugin.getManifest().getId());
+		}
+	}
+
+	public static void validate(MeshPlugin plugin) {
```
Comment (Jotschi) [inline plugin-api/src/main/java/com/gentics/mesh/plugin/util/PluginUtils.java]: Move validate into plugin registry?

**Comment by philippguertler** [comment]
When implementing both `GraphQLPlugin` and `RestPlugin`, the default implementations of `apiName` are conflicting.

**Comment by philippguertler** [comment]
When implementing the `RestPlugin` interface and using the projectRouter, how can I access the queried project?

**Comment by philippguertler** [comment]
How do I deploy my plugin when writing an integration test using the `MeshLocalServer` class rule?

**Comment by Jotschi** [comment]
> How do I deploy my plugin when writing an integration test using the `MeshLocalServer` class rule?

We should maybe add a `withPlugin` method to the MeshLocalServer.

**Pull request merged**